### PR TITLE
Openshift route fix

### LINKED ISSUE #95: CRD updater
- [x] Support patch for Cluster roles

- [x] Support patch for Services

- [x] Include RoleRef during the Cluster role binding patch method

- [x] Add binary data comparison for config map

### LINKED ISSUE #239: router and image registry urls are not getting added to private Docker registry in OpenShift 4.0
**Describe the bug**
The routes and registries are not part of default namespace in OpenShift 4.0. So the code to retrieve the default routes and image registries are failing. This is because the image registry will not be in default instead it will be in image-registry namespace and for router, it should be in openshift-ingress

**Environment (please complete the following information):**
 - Openshift Version: 4.0

**To Reproduce**
1. Create OpsSight
2. Check the logs or see the privateDockerRegistries array, the default router and image registries will not be added to it. Instead the logs shown the error message as unable to find the default router and image registry

**Expected behavior**
The default router and image registries should be added to the privateDockerRegistries

**Actual behavior**
The default router and image registries is not getting added to the privateDockerRegistries.

### LINKED ISSUE #243: expose-ui not updated
**Describe the bug**
When updating the operator exposed type, it is not updated from `NodePort` to `LoadBalancer` or vice versa

**To Reproduce**
1. Create operator without service exposed
2. Update operator to expose it to be `NodePort`
3. Update operator to expose it to be `Loadbalancer`

### LINKED ISSUE #244: Ability to unexpose operator-ui once it's exposed from synopsyctl
**Describe the bug**
User currently cannot unexpose operator-ui from synopsysctl

**Expected behavior**
`synopsysctl update operator -e=""` should delete the exposed-ui

**Actual behavior**
`synopsysctl update operator -e=""` does not delete the exposed-ui

### PR DESCRIPTION
**Pull request by msenmurugan**
* Fix all OpenShift route issues in OpsSight, Alert, Black Duck and Operator
* Closes #243 
* Closes #244 
* Closes #239 - OpenShift router and image registry service in OpsSight #239 
* Closes #95 - crud updater patch service, cluster role and role binding #95 
* add list and delete permissions to cluster role

### CODE REVIEW

**Comment by operator-r2d2** [comment]
[Keywords](https://help.github.com/articles/closing-issues-using-keywords) which can automatically close issues and at(@) mentions are not allowed in commit messages.

**The list of commits with invalid commit messages**:

- [c43ec1a](https://github.com/blackducksoftware/synopsys-operator/commits/c43ec1adcb7d4b62f646f5f4124aed20aadd933a) Fix for #243 and #244

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by operator-r2d2** [comment]
[APPROVALNOTIFIER] This PR is **NOT APPROVED**

This pull-request has been approved by:
To fully approve this pull request, please assign additional approvers.
We suggest the following additional approver: **msenmurugan**

If they are not already assigned, you can assign the PR to them by writing `/assign @msenmurugan` in a comment when ready.

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=blackducksoftware%2Fsynopsys-operator).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details open>
Needs approval from an approver in each of these files:

- **[OWNERS](https://github.com/blackducksoftware/synopsys-operator/blob/master/OWNERS)**

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

**Comment by operator-r2d2** [comment]
[Keywords](https://help.github.com/articles/closing-issues-using-keywords) which can automatically close issues and at(@) mentions are not allowed in commit messages.

**The list of commits with invalid commit messages**:

- [c43ec1a](https://github.com/blackducksoftware/synopsys-operator/commits/c43ec1adcb7d4b62f646f5f4124aed20aadd933a) Fix for #243 and #244

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

**Comment by operator-r2d2** [comment]
[Keywords](https://help.github.com/articles/closing-issues-using-keywords) which can automatically close issues and at(@) mentions are not allowed in commit messages.

**The list of commits with invalid commit messages**:

- [c43ec1a](https://github.com/blackducksoftware/synopsys-operator/commits/c43ec1adcb7d4b62f646f5f4124aed20aadd933a) Fix for #243 and #244

<details>

Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md).  If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).
</details>

File: pkg/util/common.go [hunk 1]
```diff
@@ -974,47 +975,57 @@ func IsClusterRoleRuleExist(oldRules []rbacv1.PolicyRule, newRule rbacv1.PolicyR
...
-		}
+		log.Debugf("unable to get route client")
+		return nil
 	}
 	return routeClient
 }
 
-// GetOpenShiftRoutes get a OpenShift routes
-func GetOpenShiftRoutes(routeClient *routeclient.RouteV1Client, namespace string, name string) (*routev1.Route, error) {
+// GetRoute get an OpenShift routes
```
Comment (rrati) [inline pkg/util/common.go]: gets an OpenShift route

**Comment by operator-r2d2** [comment]
[APPROVALNOTIFIER] This PR is **APPROVED**

This pull-request has been approved by: *<a href="https://github.com/blackducksoftware/synopsys-operator/pull/257#pullrequestreview-232564871" title="Approved">rrati</a>*

The full list of commands accepted by this bot can be found [here](https://go.k8s.io/bot-commands?repo=blackducksoftware%2Fsynopsys-operator).

The pull request process is described [here](https://git.k8s.io/community/contributors/guide/owners.md#the-code-review-process)

<details >
Needs approval from an approver in each of these files:

- ~~[OWNERS](https://github.com/blackducksoftware/synopsys-operator/blob/master/OWNERS)~~ [rrati]

Approvers can indicate their approval by writing `/approve` in a comment
Approvers can cancel approval by writing `/approve cancel` in a comment
</details>

File: pkg/crdupdater/helpers.go [hunk 2]
```diff
@@ -119,6 +121,39 @@ func isLabelsExist(expectedLabels map[string]label, actualLabels map[string]stri
...
+	return servicePorts
+}
+
+type policyRule struct {
+	Verbs     []string `json:"verbs" protobuf:"bytes,1,rep,name=verbs"`
+	APIGroups []string `json:"apiGroups,omitempty" protobuf:"bytes,2,rep,name=apiGroups"`
+	Resources []string `json:"resources,omitempty" protobuf:"bytes,3,rep,name=resources"`
+}
+
+func sortPolicyRule(rules []rbacv1.PolicyRule) []policyRule {
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply pkg/crdupdater/helpers.go]: Kubernetes is not sorting the policy rule while creating the resource. Whatever order the user provides, it is creating in the same order. That's why I didn't sorted the policy rules

File: pkg/api/types.go [hunk 3]
```diff
@@ -23,8 +23,20 @@ package api
...
 )
 
+// Route defines the route component
+type Route struct {
+	Namespace          string
+	Name               string
+	Kind               string
+	ServiceName        string
+	PortName           string
+	Labels             map[string]string `json:"labels,omitempty"`
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply pkg/api/types.go]: removed

File: pkg/util/common.go (hunk 1 above)
Comment (msenmurugan) [reply pkg/util/common.go]: done

File: pkg/apps/alert/latest/route.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	"fmt"
+	"strings"
+
+	"github.com/blackducksoftware/synopsys-operator/pkg/api"
+	routev1 "github.com/openshift/api/route/v1"
+)
+
+// GetOpenShiftRoute creates the OpenShift route component for the alert
+func (a *SpecConfig) GetOpenShiftRoute() *api.Route {
+	if strings.ToUpper(a.config.ExposeService) == "OPENSHIFT" {
```
[the remark this answers is not in the document]
Comment (msenmurugan) [reply pkg/apps/alert/latest/route.go]: fixed

**Comment by msenmurugan** [comment]
Created another PR with rebase

**Pull request closed without merging**